src/deck/drivers/src/aideck.c

```diff
@@ -35,62 +35,141 @@
+
+static xQueueHandle espTxQueue;
+static xQueueHandle espRxQueue;
+
+static uart_transport_packet_t espTxp;
+static uart_transport_packet_t espRxp;
+
+static EventGroupHandle_t evGroup;
+#define ESP_CTS_EVENT (1 << 0)
+#define ESP_CTR_EVENT (1 << 1)
+#define ESP_TXQ_EVENT (1 << 2)
+
+static void ESP_RX(void *param)
 {
   systemWaitStart();
-  if (espDeckFlasherCheckVersionAndBoot() == false)
+
+  while (1)
   {
-    DEBUG_PRINT("ESP32 not booted.");
-    while (1)
+    // Wait for start!
+    do
     {
-      vTaskDelay(portMAX_DELAY);
+      uart2GetDataWithTimeout(&espRxp.start, (TickType_t)portMAX_DELAY);
+    } while (espRxp.start != 0xFF);
+
+    uart2GetDataWithTimeout(&espRxp.length, (TickType_t)portMAX_DELAY);
+
+    if (espRxp.length == 0)
+    {
+      xEventGroupSetBits(evGroup, ESP_CTS_EVENT);
+    }
+    else
+    {
+      for (int i = 0; i < espRxp.length; i++)
+      {
+        uart2GetDataWithTimeout(&espRxp.data[i], (TickType_t)portMAX_DELAY);
+      }
+
+      xQueueSend(espRxQueue, &espRxp, portMAX_DELAY);
+      xEventGroupSetBits(evGroup, ESP_CTR_EVENT);
     }
   }
-#ifdef DEBUG_NINA_PRINT
+}
+
+static void ESP_TX(void *param)
+{
   systemWaitStart();
-  vTaskDelay(M2T(1000));
-  DEBUG_PRINT("Starting reading out NINA debugging messages:\n");
-  vTaskDelay(M2T(2000));
-  uart2Init(115200);
 
-  // Pull the reset button to get a clean read out of the data
-  pinMode(DECK_GPIO_IO4, OUTPUT);
-  digitalWrite(DECK_GPIO_IO4, LOW);
-  vTaskDelay(10);
-  digitalWrite(DECK_GPIO_IO4, HIGH);
-  pinMode(DECK_GPIO_IO4, INPUT_PULLUP);
+  uint8_t ctr[] = {0xFF, 0x00};
+  EventBits_t evBits = 0;
+
+  // We need to hold off here to make sure that the RX task
+  // has started up and is waiting for chars, otherwise we might send
+  // CTR and miss CTS (which means that the ESP32 will stop sending CTS
+  // too early and we cannot sync)
+  vTaskDelay(100);
+
+  // Sync with ESP32 so both are in CTS
+  do
+  {
+    uart2SendData(sizeof(ctr), (uint8_t *)&ctr);
+    vTaskDelay(100);
+    evBits = xEventGroupGetBits(evGroup);
+  } while ((evBits & ESP_CTS_EVENT) != ESP_CTS_EVENT);
 
-  // Read out the byte the NINA sends and immediately send it to the console.
-  uint8_t byte;
   while (1)
   {
-    if (uart2GetDataWithDefaultTimeout(&byte) == true)
+    // If we have nothing to send then wait, either for something to be
+    // queued or for a request to send CTR
+    if (uxQueueMessagesWaiting(espTxQueue) == 0)
     {
-      consolePutchar(byte);
+      evBits = xEventGroupWaitBits(evGroup,
+                                   ESP_CTR_EVENT | ESP_TXQ_EVENT,
+                                   pdTRUE,  // Clear bits before returning
+                                   pdFALSE, // Wait for any bit
+                                   portMAX_DELAY);
+      if ((evBits & ESP_CTR_EVENT) == ESP_CTR_EVENT)
+      {
+        uart2SendData(sizeof(ctr), (uint8_t *)&ctr);
+      }
+      if ((evBits & ESP_TXQ_EVENT) == ESP_TXQ_EVENT)
+      {
+      }
```

Comment:
What should happen here?

Reply:
Nothing, there used to be a debug printout so the flow could be followed. Removed the printouts and forgot this to remove the if. I'll update it.

Reply:
I've pushed a new commit to fix this one.